snake.asm

```diff
@@ -1,42 +1,49 @@
-	cmp [di], ch        ; compare the character at the new position of the snake to the snake character 
-	je start            ; if the result of the previous comparison yielded an equal response reset the game (the snake hit itself)
-	push di             ; push di to the stack to save the new position 
-	cmp [di], cl        ; compare the character at the new position of the snake to the food character 
-	mov [di], ch        ; move the snake character to the new position 
-	je .food            ; if the result of the last comparison yielded an equal response (the snake ate the food) print new food and don't continue to delete the snake's tail
-	es lodsw            ; use the fact si points to the stack to load the first  thing pushed to the stack (the end of the tail) and "pop" it into ax
-	xchg ax, bx         ; exchange ax and bx to save one byte and to put the position in ax in bx and to move bh (which is invisible in cp437) into ah
-	mov [bx], ah        ; move the invisible character instead of the snake character 
-	jmp SHORT .input    ; loop back to process keyboard input
+push 0xB800             ; Push the offset to video memory for DMA (Direct Memory Access)
+pop ds                  ; Load the value into the data segment register to access video memory
+; Every data access will read/write to video memory
+mov cx, 0xFA0           ; Store the screen size (80x25 characters) in text mode 3 into cx, also hold cp437 values for food and snake body, and the screen width in cl
+std                     ; Set the direction flag to reverse the operation of lodsw for FIFO processing
+start:                  ; Start new game setup
+mov ax, 0x3             ; Set ah=0 (change video mode) and al=3 (80x25 16 color text mode)
+int 0x10                ; Initiate Video BIOS interrupt
+mov di, 0x7D0           ; Set the initial position of the snake's head at the middle of the screen (0xFA0/2)
+mov si, sp              ; Setting si to the stack pointer for loading stack data using lodsw
+.food:                  ; Position new food
+imul bx, sp             ; Multiply bx by sp to pseudo-randomize bx, which points to the food position
+and bx, cx              ; Make bx divisible by 4 to fit on the screen (align with character positions). After anding, we now know that bx < cx (size of the screen) therefore it fits inside
+cmp [bx], ch            ; Compare character at memory location indicated by bx with snake body character
+je .food                ; If the comparison is true, jump to .food label to generate new food position
+mov [bx], cl            ; Place a food character at the calculated position
+.input:                 ; Process keyboard input
+in al, 0x60             ; Read keyboard input from port 0x60
+mov bx, 0x4             ; Set default distance for snake movement to 4 (one step leftwards)
+and al, 0x1E            ; Mask out non-arrow key bits (up, down, left, right)

```

Comment:
That's not what it'll do, it'll and al with 1e which result with a different parity bit whether the key was one of the sides of up/down